### PR TITLE
Inspect parents paths when checking for moves

### LINKED ISSUE #52669: "borrow of possibly uninitialized variable" instead of "use of moved value"
NLL combined with a struct field assignment gives an error mentioning the use/borrow of an uninitialized variable instead of the use/borrow of a moved variable.

# Reproduction example

*Simplified code based on @Yatekii's question on IRC*
[Playground](https://play.rust-lang.org/?gist=3458f67aa4fb164d002ad105c6d6d497&version=nightly&mode=debug&edition=2015)
```rust
#![feature(nll)]

struct A {
    b: B,
}

#[derive(Clone)]
struct B;

fn foo(_: A) {}

fn bar(mut a: A) -> B {
    // Commenting the line below gives a move error instead of uninitialized
    // error.
    a.b = B;
    foo(a);
    a.b.clone()
}

fn main() {}
```

# Expected behavior

The compiler gives an error about using a moved value, similar to the following:

Without NLL:
```
error[E0382]: use of moved value: `a.b`
  --> src/main.rs:17:5
   |
16 |     foo(a);
   |         - value moved here
17 |     a.b.clone()
   |     ^^^ value used here after move
   |
   = note: move occurs because `a` has type `A`, which does not implement the `Copy` trait
```

With NLL but without the field assignment:
```
error[E0382]: borrow of moved value: `a.b`
  --> src/main.rs:17:5
   |
16 |     foo(a);
   |         - value moved here
17 |     a.b.clone()
   |     ^^^ value borrowed here after move
   |
   = note: move occurs because `a` has type `A`, which does not implement the `Copy` trait
```

# Actual behavior

The error mentions an uninitialized value instead of a moved value:

```
error[E0381]: borrow of possibly uninitialized variable: `a.b`
  --> src/main.rs:17:5
   |
17 |     a.b.clone()
   |     ^^^ use of possibly uninitialized `a.b`
```

### PR DESCRIPTION
**Pull request by spastorino**
Closes #52669

### CODE REVIEW

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 1]
```diff
@@ -552,9 +552,17 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 // created by `StorageDead` and at the beginning
                 // of a function.
             } else {
+                let mut mpis = vec![mpi];
+
+                let mut curr_mpi = mpi;
+                while let Some(parent_mpi) = &self.move_data.move_paths[curr_mpi].parent {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/error_reporting.rs]: you probably don't need the `&` here or the `*` on the lines below

Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/error_reporting.rs]: I'd also like to see a comment explaining this. Something like:

---

If we are found a use of `a.b.c` which was in error, then we want to look for moves not only of `a.b.c` but also `a.b` and `a`.

Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/error_reporting.rs]: Writing that out makes me realize that probably there is a similar bug the other way... e.g., what happens here?

```rust
fn main() {
    let x = (vec![1, 2, 3], );
    drop(x.0);
    drop(x);
}
```

Ah, trying it on play, it seems to work ok. Interesting. I guess that the `moves` data already handles this case -- i.e., a move of `x.0` is marked as a move of `x`. Maybe add this to the comment:

---

Note that the moves data already includes "parent" paths, so we don't have to worry about the other case: that is, if there is a move of `a.b.c`, it is already marked as a move of `a.b` and `a` as well, so we will generate the correct errors there. 

---

but @pnkfelix can you verify that comment? ☝️

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks good. Left some requests for comments. :)

File: src/librustc_mir/borrow_check/error_reporting.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/error_reporting.rs]: Also, I wonder if we want to make a helper for iterating over all parents...

Comment (spastorino) [reply src/librustc_mir/borrow_check/error_reporting.rs]: @nikomatsakis everything is fixed.

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**